src/main/java/org/opensearch/ad/ratelimit/RateLimitedQueue.java

```diff
@@ -0,0 +1,465 @@
+        this.singleRequestSize = singleRequestSizeInBytes;
+        this.maxHeapPercentForQueue = maxHeapPercentForQueueSetting.get(settings);
+        this.queueSize = (int) (heapSizeInBytes * maxHeapPercentForQueue / singleRequestSizeInBytes);
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(maxHeapPercentForQueueSetting, it -> {
+            int oldQueueSize = queueSize;
+            this.queueSize = (int) (this.heapSize * maxHeapPercentForQueue / this.singleRequestSize);
+            LOG.info(new ParameterizedMessage("Queue size changed from [{}] to [{}]", oldQueueSize, queueSize));
+        });
+
+        this.queueName = queueName;
+        this.random = random;
+        this.adCircuitBreakerService = adCircuitBreakerService;
+        this.threadPool = threadPool;
+        this.maxQueuedTaskRatio = maxQueuedTaskRatio;
+        this.clock = clock;
+        this.mediumSegmentPruneRatio = mediumSegmentPruneRatio;
+        this.lowSegmentPruneRatio = lowSegmentPruneRatio;
+
+        this.lastSelectedSegmentId = null;
+        this.requestSegments = new ConcurrentSkipListMap<>();
+        this.cooldownStart = Instant.MIN;
+        this.coolDownMinutes = (int) (COOLDOWN_MINUTES.get(settings).getMinutes());
+        this.maintenanceFreqConstant = maintenanceFreqConstant;
+        this.stateTtl = stateTtl;
+        this.nodeStateManager = nodeStateManager;
+    }
+
+    protected String getQueueName() {
+        return queueName;
+    }
+
+    /**
+     * To add fairness to multiple detectors, HCAD allocates queues at a per
+     * detector granularity and pulls off requests across similar queues in a
+     * round-robin fashion.  This way, if one detector has a much higher
+     * cardinality than other detectors,  the unfinished portion of that
+     * detector’s workload times out, and other detectors’ workloads continue
+     * operating with predictable performance. For example, for loading checkpoints,
+     * HCAD pulls off 10 requests from one detector’ queues, issues a mget request
+     * to ES, wait for it to finish, and then does it again for other detectors’
+     * queues.  If one queue does not have more than 10 requests, HCAD dequeues
+     * the next batches of messages in the round-robin schedule.
+     * @return next queue to fetch requests
+     */
+    protected Optional<BlockingQueue<RequestType>> selectNextQueue() {
+        if (true == requestSegments.isEmpty()) {
+            return Optional.empty();
+        }
+
+        String startId = lastSelectedSegmentId;
+        try {
+            for (int i = 0; i < requestSegments.size(); i++) {
+                if (startId == null || requestSegments.size() == 1 || startId.equals(requestSegments.lastKey())) {
+                    startId = requestSegments.firstKey();
+                } else {
+                    startId = requestSegments.higherKey(startId);
+                }
+
+                if (startId.equals(SegmentPriority.LOW.name())) {
```

Comment:
Should we have different logic for high priority as well? We just pull high and medium requests with round-robin way?

Reply:
Yes, we just pull high and medium requests with round-robin way.  We can definitely add a special branch for high priority requests in the future :)